File: soundex/golang/soundex.go
package main

import "strings"

/**
 * Soundex algorithm implementation in Go.
 * NOTE: will panic if the given word is an empty string.
 *
 * @param {string} word - The word to convert to Soundex.
 * @returns {string} The Soundex representation of the word.
 */
func Soundex(word string) string {
	assert(len(word) > 0, "`value` must be a non-empty string")

	result := strings.ToUpper(string(word[0])) // first letter is always included
	previous := getPhonetic(word[0])           // phonetic value of the first letter

	for i := 1; i < len(word); i++ {
		c := getPhonetic(word[i])
		if c == SKIP_PHONETIC {
			previous = SKIP_PHONETIC
		} else if c != NO_VALUE && c != previous {
			result += string(c)
			if len(result) == 4 {
				return result
			}
			previous = c
		}
	}

	return (result + "000")[:4]
}

const NO_VALUE = 255
const SKIP_PHONETIC = '0'

var getPhonetic func(c byte) uint8

func init() {
	var phoneticMap = [26]uint8{
		'0', '1', '2', '3', '0', '1', '2', '0', '0', '2', '2', '4', '5', '5', '0', '1', '2', '6', '2', '3', '0', '1', '0', '2', '0', '2',
	}

	const LOOKUP_OFFSET = 65   // 'A' char code
	const UPPER_BIT_MASK = 223 // 11011111

	getPhonetic = func(c byte) uint8 {
		lookupKey := (UPPER_BIT_MASK & c) - LOOKUP_OFFSET
		if lookupKey > 25 {
			return NO_VALUE
		}

		// flip the 6th bit to make it uppercase and subtract 65 to get the index
		return phoneticMap[lookupKey]
	}
}

func assert(condition bool, message string) {
	if !condition {
		panic(message)
	}
}

File: soundex/typescript/soundex.ts
/**
 * Converts a word to its Soundex representation.
 * @param word - The word string to convert.
 * @returns The Soundex representation of the provided word.
 */
export function soundex(word: string): string {
  assert(word && word.length > 0, '`value` must be a non-empty string')

  let result = word.charAt(0).toUpperCase() // first letter is always included
  let previous = getPhonetic(word.charCodeAt(0)) // phonetic value of the first letter

  for (let i = 1; i < word.length; i++) {
    const c = getPhonetic(word.charCodeAt(i))
    if (c === SKIP_PHONETIC) {
        previous = SKIP_PHONETIC // skip phonetic value
    } else if (c !== NO_VALUE && c !== previous) {
        result += (previous = c) // append phonetic value
        if (result.length === 4) {
            return result
        }
    }
  }

  return result.padEnd(4, '0') // pad with zeros
}

const NO_VALUE = undefined
const SKIP_PHONETIC = 0
const getPhonetic = (() => {
  const m = new Uint8Array([
    0, 1, 2, 3, 0, 1, 2, 0, 0, 2, 2, 4, 5, 5, 0, 1, 2, 6, 2, 3, 0, 1, 0, 2, 0, 2
  ])

  const LOOKUP_OFFSET = 65    /* 'A' char code */
  const UPPER_BIT_MASK = 223  /* 11011111 */

  // flip the 6th bit to make it uppercase and subtract 65 to get the index
  return c => m[(UPPER_BIT_MASK & c) - LOOKUP_OFFSET]
})() as (charCode: number) => number | undefined

// deno-lint-ignore no-explicit-any
function assert(condition: any, message: string) {
  if (!condition) {
    throw new Error(message)
  }
}

